File: code/ex552.java
class Solution {
    public int checkRecord(int n) {
        long mod = (long)Math.pow(10, 9) + 7;
        
        // Define dp arrays
        long[] dp00 = new long[n + 1];  // 'A' never appeared till n, ending with no 'L'
        long[] dp01 = new long[n + 1];  // 'A' never appeared till n, ending with one 'L'
        long[] dp02 = new long[n + 1];  // 'A' never appeared till n, ending with two 'L'
        long[] dp10 = new long[n + 1];  // 'A' appeared once till n, ending with no 'L'
        long[] dp11 = new long[n + 1];  // 'A' appeared once till n, ending with one 'L'
        long[] dp12 = new long[n + 1];  // 'A' appeared once till n, ending with two 'L'

        // Init dp arrays
        dp00[0] = 1;
        
        for (int i = 1; i <= n; i++) {
            dp00[i] = (dp00[i - 1] + dp01[i - 1] + dp02[i - 1]) % mod;
            dp01[i] = dp00[i - 1];
            dp02[i] = dp01[i - 1];
            
            dp10[i] = (dp00[i - 1] + dp01[i - 1] + dp02[i - 1] + dp10[i - 1] + dp11[i - 1] + dp12[i - 1]) % mod;
            dp11[i] = dp10[i - 1];
            dp12[i] = dp11[i - 1];
        }
        
        return (int)((dp00[n] + dp01[n] + dp02[n] + dp10[n] + dp11[n] + dp12[n]) % mod);
    }
}

File: code/ex552.cpp
const int mod = 1e9 + 7;

class Solution {
public:
    int checkRecord(int n) {
        if (n == 0)    return 1;
        
        // With As before
        long dp1 = 0;    // End with no L
        long dp2 = 0;    // End with one L
        long dp3 = 0;    // End with two continuous Ls
        
        // With no A before
        long dp4 = 1;    // End with no L
        long dp5 = 0;    // End with one L
        long dp6 = 0;    // End with two continuous Ls
        
        for (int i = 1; i <= n; i++) {
            long p1 = dp1, p2 = dp2, p3 = dp3, p4 = dp4, p5 = dp5, p6 = dp6;
            dp1 = (p1 + p2 + p3 + p4 + p5 + p6) % mod;
            dp2 = p1;
            dp3 = p2;
            dp4 = (p4 + p5 + p6) % mod;
            dp5 = p4;
            dp6 = p5;
        }
        
        return (dp1 + dp2 + dp3 + dp4 + dp5 + dp6) % mod;
    }
};